### PR TITLE
runtime light

### PR DESCRIPTION
**Pull request by astrophysik**

### CODE REVIEW

File: runtime-light/allocator/runtime-light-allocator.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    ptr = rt_ctx.runtime_allocator.memory_resource.allocate(size);
+  }
+  return ptr;
+}
+
+void *RuntimeAllocator::alloc0_script_memory(size_t size) noexcept {
+  php_assert(size);
+  if (unlikely(!is_script_allocator_available())) {
+    php_assert(false && "need to add zeroing mem here");
+    return get_platform_context()->allocator.alloc(size);
```
Comment (apolyakov) [inline runtime-light/allocator/runtime-light-allocator.cpp:73]: You forgot  zeroing of the allocated memory

File: runtime-light/component/component.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <coroutine>
+#include <csetjmp>
+#include <queue>
+#include <functional>
+#include <setjmp.h>
```
Comment (apolyakov) [inline runtime-light/component/component.h:7]: `<csetjmp>?`

File: runtime-light/component/component.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#include "runtime-light/utils/context.h"
+
+struct ComponentState {
+  template<typename Key, typename Value>
+  using unordered_map = memory_resource::stl::unordered_map<Key, Value, memory_resource::unsynchronized_pool_resource>;
+  template<typename T>
+  using deque = memory_resource::stl::deque<T, memory_resource::unsynchronized_pool_resource>;
+
+  ComponentState() :
+    runtime_allocator(16 * 1024u, 0),
```
Comment (apolyakov) [inline runtime-light/component/component.h:27]: Let's not use magic constants

File: runtime-light/core/globals/php-script-globals.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+#include "runtime-light/component/component.h"
+
+PhpScriptMutableGlobals &PhpScriptMutableGlobals::current() noexcept {
+  return get_component_context()->php_script_mutable_globals_singleton;
+}
+
+void PhpScriptMutableGlobals::once_alloc_linear_mem(unsigned int n_bytes) {
+  php_assert(g_linear_mem == nullptr);
+  g_linear_mem = static_cast<char *>(get_platform_context()->allocator.alloc(n_bytes));
```
Comment (apolyakov) [inline runtime-light/core/globals/php-script-globals.cpp:15]: Why not `alloc0`?

File: runtime-light/core/globals/php-script-globals.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  memset(g_linear_mem, 0, n_bytes);
+}
+
+void PhpScriptMutableGlobals::once_alloc_linear_mem(const char *lib_name, unsigned int n_bytes) {
+  int64_t key_lib_name = string_hash(lib_name, strlen(lib_name));
+  php_assert(libs_linear_mem.find(key_lib_name) == libs_linear_mem.end());
+  libs_linear_mem[key_lib_name] = static_cast<char *>(get_platform_context()->allocator.alloc(n_bytes));
```
Comment (apolyakov) [inline runtime-light/core/globals/php-script-globals.cpp:22]: Same

File: runtime-light/core/kphp-core-impl/kphp-core-context.cpp [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#include "runtime-core/runtime-core.h"
+#include "runtime-light/utils/context.h"
+#include "runtime-light/component/component.h"
+
+KphpCoreContext &KphpCoreContext::current() noexcept {
+  return get_component_context()->kphp_core_context;
+}
+
+void KphpCoreContext::init() {
+  init_string_buffer_lib(1024, (1 << 24));
```
Comment (apolyakov) [inline runtime-light/core/kphp-core-impl/kphp-core-context.cpp:10]: Let's not use magic constants

File: runtime-light/stdlib/output-control.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  httpResponse.output_buffers[httpResponse.current_buffer].clean();
+}
+
+bool f$ob_end_clean() {
+  Response &httpResponse = get_component_context()->response;
+  if (httpResponse.current_buffer == system_level_buffer) {
+    return false;
+  }
+
+  httpResponse.current_buffer--;
```
Comment (apolyakov) [inline runtime-light/stdlib/output-control.cpp:19]: `--httpResponse.current_buffer`

File: runtime-light/stdlib/superglobals.h [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include "runtime-core/runtime-core.h"
+#include "runtime-light/coroutine/task.h"
+
+enum QueryType {
```
Comment (apolyakov) [inline runtime-light/stdlib/superglobals.h:6]: Let's use `enum class` instead of raw enums

File: runtime-light/streams/interface.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * component query non blocked interface
+ * */
+class_instance<C$ComponentStream> f$component_open_stream(const string &name);
+task_t<class_instance<C$ComponentStream>> f$component_accept_stream();
+
+int64_t f$component_stream_write_nonblock(const class_instance<C$ComponentStream> & stream, const string & message);
+string f$component_stream_read_nonblock(const class_instance<C$ComponentStream> & stream);
+
+task_t<int64_t> f$component_stream_write_exact(const class_instance<C$ComponentStream> & stream, const string & message);
```
Comment (apolyakov) [inline runtime-light/streams/interface.h:31]: It should not be marked as `non blocking`

File: runtime-light/streams/interface.h [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+task_t<string> f$component_stream_read_exact(const class_instance<C$ComponentStream> & stream, int64_t len);
```
Comment (apolyakov) [inline runtime-light/streams/interface.h:32]: Same

File: runtime-light/streams/streams.h [hunk 11]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <utility>
+#include <cstdint>
+
+#include "runtime-core/runtime-core.h"
+
+#include "runtime-light/coroutine/task.h"
+
+enum StreamRuntimeStatus {
```
Comment (apolyakov) [inline runtime-light/streams/streams.h:10]: `enum class`?

**Review by apolyakov — COMMENTED** [review]
This PR lacks of copyright message at the start of lots of added files.

File: runtime-light/core/globals/php-script-globals.cpp (hunk 4 above)
Comment (astrophysik) [reply runtime-light/core/globals/php-script-globals.cpp:15]: platform allocator does not have alloc0

File: runtime-light/allocator/runtime-light-allocator.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+#include "runtime-light/utils/panic.h"
+
+namespace {
+bool is_script_allocator_available() {
+  return get_component_context() != nullptr;
+}
+
+void request_extra_memory() {
+  ComponentState &rt_ctx = *get_component_context();
+  // todo:k2 make extra mem size dynamic
```
Comment (apolyakov) [inline runtime-light/allocator/runtime-light-allocator.cpp:16]: Can we get it work?

**Review by apolyakov — CHANGES_REQUESTED** [review]

File: runtime-light/core/globals/php-script-globals.cpp (hunk 4 above)
Comment (apolyakov) [reply runtime-light/core/globals/php-script-globals.cpp:15]: Add `alloc0` to `RuntimeAllocator::global_memory_alloc`?

File: runtime-light/allocator/runtime-light-allocator.cpp [hunk 13]
```diff
@@ -6,25 +6,26 @@
 #include "runtime-light/component/component.h"
 #include "runtime-light/utils/panic.h"
 
+static constexpr size_t minimum_request_extra_memory_size = 16 * 1024u;
```
Comment (apolyakov) [inline runtime-light/allocator/runtime-light-allocator.cpp:9]: We use caps + snake case for globals, aren't we?

File: runtime-light/allocator/runtime-light-allocator.cpp [hunk 14]
```diff
@@ -54,50 +54,51 @@ void RuntimeAllocator::free() {
...
+    return alloc_global_memory(size);
   }
 
   ComponentState &rt_ctx = *get_component_context();
 
   void *ptr = rt_ctx.runtime_allocator.memory_resource.allocate(size);
-  if (ptr == nullptr) {
-    request_extra_memory(size);
+  size_t previous_multiplier = 1;
+  while (ptr != nullptr) {
```
Comment (apolyakov) [inline runtime-light/allocator/runtime-light-allocator.cpp:64]: Shouldn't we have a fallback here? What if `memory_resource.allocate` never returns a non-`nullptr` value?

File: runtime-light/core/globals/php-script-globals.cpp [hunk 15]
```diff
@@ -12,14 +12,14 @@ PhpScriptMutableGlobals &PhpScriptMutableGlobals::current() noexcept {
 
 void PhpScriptMutableGlobals::once_alloc_linear_mem(unsigned int n_bytes) {
   php_assert(g_linear_mem == nullptr);
-  g_linear_mem = static_cast<char *>(get_platform_context()->allocator.alloc(n_bytes));
+  g_linear_mem = static_cast<char *>(RuntimeAllocator::current().alloc0_global_memory(n_bytes));
   memset(g_linear_mem, 0, n_bytes);
```
Comment (apolyakov) [inline runtime-light/core/globals/php-script-globals.cpp:16]: Now it's redundant

File: runtime-light/core/globals/php-script-globals.cpp [hunk 16]
```diff
@@ -12,14 +12,14 @@ PhpScriptMutableGlobals &PhpScriptMutableGlobals::current() noexcept {
...
 }
 
 void PhpScriptMutableGlobals::once_alloc_linear_mem(const char *lib_name, unsigned int n_bytes) {
   int64_t key_lib_name = string_hash(lib_name, strlen(lib_name));
   php_assert(libs_linear_mem.find(key_lib_name) == libs_linear_mem.end());
-  libs_linear_mem[key_lib_name] = static_cast<char *>(get_platform_context()->allocator.alloc(n_bytes));
+  libs_linear_mem[key_lib_name] = static_cast<char *>(RuntimeAllocator::current().alloc0_global_memory(n_bytes));
   memset(libs_linear_mem[key_lib_name], 0, n_bytes);
```
Comment (apolyakov) [inline runtime-light/core/globals/php-script-globals.cpp:23]: Same

File: runtime-light/streams/interface.h [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+ * component query server blocked interface
+ * */
+task_t<string> f$component_server_get_query();
+task_t<void> f$component_server_send_result(const string &message);
+
+/**
+ * component query non blocked interface
+ * */
+class_instance<C$ComponentStream> f$component_open_stream(const string &name);
+task_t<class_instance<C$ComponentStream>> f$component_accept_stream();
```
Comment (apolyakov) [inline runtime-light/streams/interface.h:30]: It's a blocking function

File: runtime-light/streams/interface.h [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+
+/**
+ * component query blocked interface
+ * */
+int64_t f$component_stream_write_nonblock(const class_instance<C$ComponentStream> &stream, const string &message);
```
Comment (apolyakov) [inline runtime-light/streams/interface.h:36]: These two function are not blocking

File: runtime-light/streams/interface.h [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+string f$component_stream_read_nonblock(const class_instance<C$ComponentStream> &stream);
+
+task_t<int64_t> f$component_stream_write_exact(const class_instance<C$ComponentStream> &stream, const string &message);
```
Comment (apolyakov) [inline runtime-light/streams/interface.h:39]: These are blocking

**Review by apolyakov — APPROVED** [review]
LGTM

**Review by apolyakov — APPROVED** [review]

**Review by apolyakov — APPROVED** [review]

**Pull request merged**